packages/wonder-blocks-core/components/clickable-behavior.js

```diff
@@ -269,12 +283,43 @@ export default class ClickableBehavior extends React.Component<
         this.dragging = false;
     }
 
-    handleClick = (e: SyntheticMouseEvent<>) => {
+    handleClick = async (e: SyntheticMouseEvent<>) => {
+        const {
+            onClick,
+            beforeNav,
+            safeWithNav,
+            skipClientNav,
+            history,
+        } = this.props;
+
         if (this.enterClick) {
             return;
-        } else if (this.props.onClick) {
+        } else if (onClick) {
             this.waitingForClick = false;
-            this.props.onClick(e);
+            onClick(e);
+        } else if (beforeNav) {
```

Comment:
I think it would be better if `onClick`, `beforeNav`, and `safeWithNav` weren't all mutually exclusive. Instead I think it would be useful if they would all be run, in that order. As the ADR says, "If both beforeNav and safeWithNav functions are passed into the component, then beforeNav will be run first, then safeWithNav."

I think supporting all the props gives users a clear way to divide up their code. For example, if I want to update local component state, do a GraphQL mutation, and start an analytics request, I could do `setState` in the `onClick` function, do the mutation in `beforeNav`, and do the analytics request in `safeWithNav`. Then, if I ever don't need one piece of that in the future, I can just remove one of the handlers, and everything else stays the same. If I have to combine them into one handler, then it could be more complicated to remove or add things in the future, since I might have to move the handler to a different prop. Supporting multiple props could also make the user experience better, since client-side navigation could wait for some server requests to finish, but start while others are still in flight. That wouldn't be possible if the user has to manage all requests themselves in a single handler.

The only complicated thing is what to do if both `beforeNav` and `safeWithNav` return promises. I think the best thing would be to run both functions, then wait for the `beforeNav` promise to resolve, then navigate (unless a server-side navigation is going to happen, when we'd have to wait for both promises to resolve).

Reply:
I can see the utility of `beforeNav` and `safeWithNav` being allowed at the same time.  Allowing `onClick` I think will lead to confusion about when to use `onClick` vs `safeWithNav` since they're almost the same thing.

> The only complicated thing is what to do if both beforeNav and safeWithNav return promises. I think the best thing would be to run both functions, then wait for the beforeNav promise to resolve, then navigate (unless a server-side navigation is going to happen, when we'd have to wait for both promises to resolve).

That makes sense.

Reply:
We chatted on Slack and agreed to allow all three props at once with `onClick` being used for synchronous code and the other two being used for aysnc code.